# 1. Data Analysis

In [39]:
using Pkg
using Plots
using Random
using GLM
using DataFrames
using LaTeXStrings
using Statistics
using StatsBase
using CSV

1.1. Import the data set. Make sure the column names are imported as intended.

In [40]:
data = CSV.read(download("https://raw.githubusercontent.com/d2cml-ai/CausalAI-Course/main/data/wage2015_subsample_inference.csv"), DataFrame);
first(data,5)

5×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg    ⋯
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float6 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │       10   9.61538  2.26336      1.0      0.0      0.0      0.0      1. ⋯
   2 │       12  48.0769   3.8728       0.0      0.0      0.0      0.0      1.
   3 │       15  11.0577   2.40313      0.0      0.0      1.0      0.0      0.
   4 │       18  13.9423   2.63493      1.0      0.0      0.0      0.0      0.
   5 │       19  28.8462   3.36198      1.0      0.0      0.0      0.0      1. ⋯
                                                              14 columns omitted

1.2. Are there missing values? Display the number of missings (if any) of each variable in the data set.

In [41]:
# Obtener el número de valores faltantes por columna usando describe
missing_counts = describe(data).nmissing
missing_df = DataFrame(Column = names(data), MissingCount = missing_counts)
println(missing_df)
# Verificar si hay o no valores faltantes
if all(count -> count == 0, missing_counts)
    println("There are no missing values")
else
    println("There are missing values")
end

21×2 DataFrame
 Row │ Column    MissingCount 
     │ String    Int64        
─────┼────────────────────────
   1 │ rownames             0
   2 │ wage                 0
   3 │ lwage                0
   4 │ sex                  0
   5 │ shs                  0
   6 │ hsg                  0
   7 │ scl                  0
   8 │ clg                  0
   9 │ ad                   0
  10 │ mw                   0
  11 │ so                   0
  12 │ we                   0
  13 │ ne                   0
  14 │ exp1                 0
  15 │ exp2                 0
  16 │ exp3                 0
  17 │ exp4                 0
  18 │ occ                  0
  19 │ occ2                 0
  20 │ ind                  0
  21 │ ind2                 0
There are no missing values


1.3. Report descriptive statistics of the variables (mean, standard deviation, percentiles, etc.). Interpret your results.

In [42]:
describe(data)

21×7 DataFrame
 Row │ variable  mean          min        median       max          nmissing   ⋯
     │ Symbol    Float64       Real       Float64      Real         Int64      ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ rownames  15636.3        10        15260.0       32643              0   ⋯
   2 │ wage         23.4104      3.02198     19.2308      528.846          0
   3 │ lwage         2.97079     1.10591      2.95651       6.2707         0
   4 │ sex           0.444466    0.0          0.0           1.0            0
   5 │ shs           0.023301    0.0          0.0           1.0            0   ⋯
   6 │ hsg           0.243883    0.0          0.0           1.0            0
   7 │ scl           0.278058    0.0          0.0           1.0            0
   8 │ clg           0.31767     0.0          0.0           1.0            0
   9 │ ad            0.137087    0.0          0.0           1.0            0   ⋯
  10 │ mw            0.259612    0.0          0.0           1.0            0
  11 │ so            0.296505    0.0          0.0           1.0            0
  12 │ we            0.216117    0.0          0.0           1.0            0
  13 │ ne            0.227767    0.0          0.0           1.0            0   ⋯
  14 │ exp1         13.7606      0.0         10.0          47.0            0
  15 │ exp2          3.01893     0.0          1.0          22.09           0
  16 │ exp3          8.23587     0.0          1.0         103.823          0
  17 │ exp4         25.118       0.0          1.0         487.968          0   ⋯
  18 │ occ        5310.74       10.0       4040.0      100000.0            0
  19 │ occ2         11.6709      1           13.0          22              0
  20 │ ind        6629.15      370.0       7370.0      100000.0            0
  21 │ ind2         13.3169      2           14.0          22              0   ⋯
                                                                1 column omitted

In [43]:
# Se crea un DataFrame con los nombres de las columnas y calcular percentiles
percentiles_df = DataFrame(Column = String[], p10 = Float64[], p25 = Float64[], p50 = Float64[], p75 = Float64[], p90 = Float64[])

# Iterando sobre cada columna numérica del DataFrame
for (col_name, col_data) in zip(names(data), eachcol(data))
    # Se comprueba si la columna es numérica
    if eltype(col_data) <: Number
        # Se calcula los percentiles específicos
        p10 = quantile(col_data, 0.10)
        p25 = quantile(col_data, 0.25)
        p50 = quantile(col_data, 0.50)
        p75 = quantile(col_data, 0.75)
        p90 = quantile(col_data, 0.90)
        
        # se agregan los resultados al DataFrame de percentiles
        push!(percentiles_df, (Column = col_name, p10 = p10, p25 = p25, p50 = p50, p75 = p75, p90 = p90))
    end
end

# Se imprimen los percentiles calculados
println(percentiles_df)

21×6 DataFrame
 Row │ Column    p10         p25         p50          p75          p90         
     │ String    Float64     Float64     Float64      Float64      Float64     
─────┼─────────────────────────────────────────────────────────────────────────
   1 │ rownames  2692.9      7132.5      15260.0      24337.8      29770.3
   2 │ wage         9.61538    13.4615      19.2308      27.7778      39.6154
   3 │ lwage        2.26336     2.59984      2.95651      3.32424      3.67922
   4 │ sex          0.0         0.0          0.0          1.0          1.0
   5 │ shs          0.0         0.0          0.0          0.0          0.0
   6 │ hsg          0.0         0.0          0.0          0.0          1.0
   7 │ scl          0.0         0.0          0.0          1.0          1.0
   8 │ clg          0.0         0.0          0.0          1.0          1.0
   9 │ ad           0.0         0.0          0.0          0.0          1.0
  10 │ mw           0.0         0.0          0.0          1.0  

Desviación estandar

In [44]:
# DataFrame con los nombres de las columnas y la desviación estándar
std_df = DataFrame(Column = String[], StdDev = Float64[])

# Iteración sobre cada columna del DataFrame
for (col_name, col_data) in zip(names(data), eachcol(data))
    # Comprobando que la columna sea numérica
    if eltype(col_data) <: Number
        # Calculo de la desviación estándar de la columna
        std_dev = std(skipmissing(col_data))
        
        # Resultados al DataFrame
        push!(std_df, (Column = col_name, StdDev = std_dev))
    end
end
println(std_df)

21×2 DataFrame
 Row │ Column    StdDev       
     │ String    Float64      
─────┼────────────────────────
   1 │ rownames   9700.74
   2 │ wage         21.003
   3 │ lwage         0.570385
   4 │ sex           0.496955
   5 │ shs           0.150872
   6 │ hsg           0.429465
   7 │ scl           0.448086
   8 │ clg           0.465616
   9 │ ad            0.343973
  10 │ mw            0.438464
  11 │ so            0.456761
  12 │ we            0.411635
  13 │ ne            0.419432
  14 │ exp1         10.6095
  15 │ exp2          4.0009
  16 │ exp3         14.489
  17 │ exp4         53.5302
  18 │ occ       11874.4
  19 │ occ2          6.96668
  20 │ ind        5333.44
  21 │ ind2          5.70102


1.4. How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [45]:
# Se filtra y cuenta a las mujeres con título universitario o superior en el 25% más rico de la muestra
rich_ed_women = filter(row -> (row.sex == 1) && ((row.clg == 1) || (row.ad == 1)) && (row.wage >= quantile(skipmissing(data.wage), 0.75)), data)

# Impresión 
println("Número de mujeres educadas en el 25% más rico: ", nrow(rich_ed_women))
println("DataFrame de mujeres educadas en el 25% más rico:")
println(rich_ed_women)

Número de mujeres educadas en el 25% más rico: 419
DataFrame de mujeres educadas en el 25% más rico:
419×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg      ad       mw       so       we       ne       exp1     exp2     exp3     exp4      occ       occ2   ind       ind2  
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64   Float64   Int64  Float64   Int64 
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │       19   28.8462  3.36198      1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0      0.0      1.0     22.0     4.84   10.648   23.4256    2015.0      6    9470.0     22
   2 │      191   42.3077  3.74497      1.0      0.0      0.0      0.0      0.0      1.0      0

1.5. How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [46]:
# Se filtra y cuenta a los hombres con título de escuela secundaria o menor en el 25% más rico de la muestra
rich_hsg_men = filter(row -> (row.sex == 0) &&  # Filtra hombres
                             ((row.hsg == 1) || (row.hsg == 1)) &&  # Filtro por nivel educativo (secundaria o menor)
                             (row.wage >= quantile(skipmissing(data.wage), 0.75)), data)  # Filtro por salario en el 25% más alto
# Impresión 
println("Número de hombres con educación secundaria o menor en el 25% más rico: ", nrow(rich_hsg_men))
println("DataFrame de hombres con educación secundaria o menor en el 25% más rico:")
println(rich_hsg_men)

Número de hombres con educación secundaria o menor en el 25% más rico: 110
DataFrame de hombres con educación secundaria o menor en el 25% más rico:
110×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg      ad       mw       so       we       ne       exp1     exp2     exp3     exp4      occ       occ2   ind      ind2  
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64   Float64   Int64  Float64  Int64 
─────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │      113   27.8846  3.32808      0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0      0.0      1.0     16.0     2.56    4.096    6.5536    6355.0     19    770.0      4
   2 │      276   28.8462  3.36198      0.0      0.

1.6. Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variables.

In [47]:
# DataFrame para 'lwage' que es el log(wage)
df_lwage = data[:, [:lwage]]

# DataFrame para el resto de las columnas a excepción de 'rownames' 'wage' y 'lwage'
data2 = select(data, Not([:rownames, :wage, :lwage]))

display(df_lwage)
display(data2)

5150×1 DataFrame
  Row │ lwage   
      │ Float64 
──────┼─────────
    1 │ 2.26336
    2 │ 3.8728
    3 │ 2.40313
    4 │ 2.63493
    5 │ 3.36198
    6 │ 2.46222
    7 │ 2.95651
    8 │ 2.95651
    9 │ 2.48491
   10 │ 2.95651
   11 │ 2.85115
  ⋮   │    ⋮
 5141 │ 3.81874
 5142 │ 3.11778
 5143 │ 2.82298
 5144 │ 3.17966
 5145 │ 2.62801
 5146 │ 2.69255
 5147 │ 3.13883
 5148 │ 3.64966
 5149 │ 3.49551
 5150 │ 2.85115
5129 rows omitted

5150×18 DataFrame
  Row │ sex      shs      hsg      scl      clg      ad       mw       so      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │     1.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5142 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
 5143 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 5144 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5145 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5146 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 5147 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      0.0
 5148 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
 5149 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 5150 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0 ⋯
                                                10 columns and 5129 rows omitted

# 2. Data wrangling

2.7. Make an array for our Y variable, which will be the logarithm of wage (lwage column)

In [48]:
#Le cambiamos de nombre a la target a df_lwage puesto que ya existe
 Y = df_lwage
first(Y,5)

5×1 DataFrame
 Row │ lwage   
     │ Float64 
─────┼─────────
   1 │ 2.26336
   2 │ 3.8728
   3 │ 2.40313
   4 │ 2.63493
   5 │ 3.36198

2.8. Make three new arrays for our predictors:

8.1. The basic model will have the columns sex hsg scl clg ad so we ne exp1 occ2 ind2. Make sure to convert occ2 and ind2 to dummies and to drop the first dummy value to prevent multicolinearity. This means:

Creación de la primera dummy occ2 (Ocupación)

In [49]:
# Número de valores a transformar en dummys
println(sort(unique(data.occ2)))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]


In [50]:
# Se crea un bucle para poder transformar los valores únicos de la variable occ2 en variables con valores de 0 y 1. 
# La denominación 1 aparecerá cuando la variable coincida con el valor para cada categoría. 
data3 = data2
for i in 2:22
    col = Symbol("occ2_$i")
    data3 = transform(data3, :occ2 => ByRow(x -> x == i ? 1 : 0) => col)
end

println(data3)

Excessive output truncated after 1841029 bytes.

Creación de la segunda dummy ind2 (codigo de insdustria)

In [51]:
# Número de valores a transformar en dummys
println(sort(unique(data2.ind2)))

[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]


In [52]:
# Se crea un bucle para poder transformar los valores únicos de la variable ind2 en variables con valores de 0 y 1. 
# La denominación 1 aparecerá cuando la variable coincida con el valor para cada categoría. 
for i in 2:21
    col = Symbol("ind2_$i")
    data3 = transform(data3, :ind2 => ByRow(x -> x == i ? 1 : 0) => col)
end

println(data3)

Excessive output truncated after 2728029 bytes.

In [53]:
# se crea el modelo básico según las especificaciones, se tiene que eliminar las variables que no son requeridas. 

modelo_basico = select!(data3, Not(["shs", "mw", "occ2", "ind2", "exp2", "exp3", "exp4"]))
modelo_basico

5150×52 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    9 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0 ⋯
   10 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
   11 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5142 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0 ⋯
 5143 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
                                                44 columns and 5129 rows omitted

8.2. The flexible model will have the same columns, and will also include polynomials for experience (exp2 exp3 exp4), as well as the interactions between all experience variables and other variables except for sex. This means

In [54]:
m_flexible = modelo_basico

insertcols!(m_flexible, :exp2 => data2.exp2, :exp3 => data2.exp3, :exp4 => data2.exp4)

m_flexible


5150×55 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    9 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0 ⋯
   10 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
   11 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5142 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0 ⋯
 5143 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
                                                47 columns and 5129 rows omitted

In [55]:
evars = [:exp1, :exp2, :exp3, :exp4]

vars = [
    :hsg, :scl, :clg, :ad, :so, :we, :ne,   
    :occ2_2, :occ2_3, :occ2_4, :occ2_5, :occ2_6, :occ2_7, :occ2_8, :occ2_9, 
    :occ2_10, :occ2_11, :occ2_12, :occ2_13, :occ2_14, :occ2_15, :occ2_16, 
    :occ2_17, :occ2_18, :occ2_19, :occ2_20, :occ2_21, :occ2_22,
    :ind2_2, :ind2_3, :ind2_4, :ind2_5, :ind2_6, :ind2_7, :ind2_8, :ind2_9,
    :ind2_10, :ind2_11, :ind2_12, :ind2_13, :ind2_14, :ind2_15, :ind2_16,
    :ind2_17, :ind2_18, :ind2_19, :ind2_20, :ind2_21
]

# Create interactions
for exp in evars
    for var in vars
        # New column name
        interac = Symbol("$(exp)_x_$(var)")
        
        m_flexible[!, interac] = m_flexible[!, exp] .* m_flexible[!, var]
    end
end

m_flexible

5150×247 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    9 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0 ⋯
   10 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
   11 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5142 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0 ⋯
 5143 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
                                               239 columns and 5129 rows omitted

8.3. The extra-flexible model will include all two-way interactions between variables, except for sex. This means

In [56]:
m_flexible2 = m_flexible


# Variables without "sex" 
vars2 = [
    :exp1, :exp2, :exp3, :exp4, :hsg, :scl, :clg, :ad, :so, :we, :ne,
    :occ2_2, :occ2_3, :occ2_4, :occ2_5, :occ2_6, :occ2_7, :occ2_8, 
    :occ2_9, :occ2_10, :occ2_11, :occ2_12, :occ2_13, :occ2_14, :occ2_15, :occ2_16, 
    :occ2_17, :occ2_18, :occ2_19, :occ2_20, :occ2_21, :occ2_22, 
    :ind2_2, :ind2_3, :ind2_4, :ind2_5, :ind2_6, :ind2_7, :ind2_8, :ind2_9, 
    :ind2_10, :ind2_11, :ind2_12, :ind2_13, :ind2_14, :ind2_15, :ind2_16, 
    :ind2_17, :ind2_18, :ind2_19, :ind2_20, :ind2_21
]

# Create two way interactions
for i in 1:length(vars2)
    for j in (i+1):length(vars2)  # begin from `i+1` to avoid duplicates
        var1 = vars2[i]
        var2 = vars2[j]

        # Name of new interactions
        interac2 = Symbol("$(var1)_x_$(var2)")

        # Create interactions
        m_flexible2[!, interac2] = m_flexible2[!, var1] .* m_flexible2[!, var2]
    end
end

m_flexible2


5150×1381 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    9 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0 ⋯
   10 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
   11 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5142 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0 ⋯
 5143 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
                                              1373 columns and 5129 rows omitted

## 3. Linear Regressions

9. Split each of the dataframes created (basic, flexible and extra-flexible models) into a training sample (80% of the data) and a test sample.

In [57]:
# Combinar el target 'Y' con el DataFrame de características del modelo básico
data_basico = hcat(Y, modelo_basico)  
rename!(data_basico, 1 => :lwage) 

# Combinar el target 'Y' con el DataFrame de características del modelo flexible
data_flexible = hcat(Y, m_flexible)  
rename!(data_flexible, 1 => :lwage) 

# Combinar el target 'Y' con el DataFrame de características del modelo extra flexible
data_extra_flexible = hcat(Y, m_flexible2)  
rename!(data_extra_flexible, 1 => :lwage)  

# Mostrar los DataFrames resultantes
println("DataFrame del modelo básico con el target:")
display(data_basico)

println("DataFrame del modelo flexible con el target:")
display(data_flexible)

println("DataFrame del modelo extra flexible con el target:")
display(data_extra_flexible)

DataFrame del modelo básico con el target:


5150×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.26336      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │ 3.8728       0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │ 2.40313      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │ 2.63493      1.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │ 3.36198      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │ 2.46222      1.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │ 2.95651      1.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │ 2.95651      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │ 2.48491      1.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 2.95651      1.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │ 2.85115      1.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │ 3.81874      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5142 │ 3.11778      0.0      1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 5143 │ 2.82298      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5144 │ 3.17966      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5145 │ 2.62801      0.0      0.0      0.0      1.0      0.0      0.0      1.0
 5146 │ 2.69255      0.0      0.0      0.0      1.0      0.0      0.0      1.0 ⋯
 5147 │ 3.13883      1.0      0.0      1.0      0.0      0.0      0.0      1.0
 5148 │ 3.64966      0.0      0.0      0.0      0.0      1.0      0.0      1.0
 5149 │ 3.49551      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5150 │ 2.85115      0.0      0.0      0.0      0.0      1.0      0.0      1.0 ⋯
                                              1374 columns and 5129 rows omitted

DataFrame del modelo flexible con el target:


5150×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.26336      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │ 3.8728       0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │ 2.40313      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │ 2.63493      1.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │ 3.36198      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │ 2.46222      1.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │ 2.95651      1.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │ 2.95651      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │ 2.48491      1.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 2.95651      1.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │ 2.85115      1.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │ 3.81874      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5142 │ 3.11778      0.0      1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 5143 │ 2.82298      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5144 │ 3.17966      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5145 │ 2.62801      0.0      0.0      0.0      1.0      0.0      0.0      1.0
 5146 │ 2.69255      0.0      0.0      0.0      1.0      0.0      0.0      1.0 ⋯
 5147 │ 3.13883      1.0      0.0      1.0      0.0      0.0      0.0      1.0
 5148 │ 3.64966      0.0      0.0      0.0      0.0      1.0      0.0      1.0
 5149 │ 3.49551      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5150 │ 2.85115      0.0      0.0      0.0      0.0      1.0      0.0      1.0 ⋯
                                              1374 columns and 5129 rows omitted

DataFrame del modelo extra flexible con el target:


5150×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.26336      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │ 3.8728       0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │ 2.40313      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │ 2.63493      1.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │ 3.36198      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │ 2.46222      1.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │ 2.95651      1.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │ 2.95651      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │ 2.48491      1.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 2.95651      1.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │ 2.85115      1.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │ 3.81874      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5142 │ 3.11778      0.0      1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 5143 │ 2.82298      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5144 │ 3.17966      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 5145 │ 2.62801      0.0      0.0      0.0      1.0      0.0      0.0      1.0
 5146 │ 2.69255      0.0      0.0      0.0      1.0      0.0      0.0      1.0 ⋯
 5147 │ 3.13883      1.0      0.0      1.0      0.0      0.0      0.0      1.0
 5148 │ 3.64966      0.0      0.0      0.0      0.0      1.0      0.0      1.0
 5149 │ 3.49551      0.0      1.0      0.0      0.0      0.0      0.0      1.0
 5150 │ 2.85115      0.0      0.0      0.0      0.0      1.0      0.0      1.0 ⋯
                                              1374 columns and 5129 rows omitted

In [58]:
using DataFrames, Random

# Fijar la semilla para la reproducibilidad de los resultados
Random.seed!(5)

# Definir una función para dividir un DataFrame en conjunto de entrenamiento y prueba
function split_train_test(df, train_ratio=0.8)
    # Mezclar aleatoriamente las filas del DataFrame
    shuffled_df = shuffle(df)
    
    # Calcular el tamaño del conjunto de entrenamiento
    n_train = round(Int, size(df, 1) * train_ratio)
    
    # Dividir el DataFrame en conjunto de entrenamiento y conjunto de prueba
    train_df = shuffled_df[1:n_train, :]
    test_df = shuffled_df[(n_train+1):end, :]
    
    return train_df, test_df
end

# Aplicar la función de división a cada DataFrame
train_basico, test_basico = split_train_test(data_basico)
train_flexible, test_flexible = split_train_test(data_flexible)
train_extra_flexible, test_extra_flexible = split_train_test(data_extra_flexible)

# Mostrar los resultados
println("Modelo Básico - Entrenamiento:")
display(train_basico)

println("Modelo Básico - Prueba:")
display(test_basico)

println("Modelo Flexible - Entrenamiento:")
display(train_flexible)

println("Modelo Flexible - Prueba:")
display(test_flexible)

println("Modelo Extra Flexible - Entrenamiento:")
display(train_extra_flexible)

println("Modelo Extra Flexible - Prueba:")
display(test_extra_flexible)

Modelo Básico - Entrenamiento:


4120×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.66883      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
    2 │ 2.59984      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    3 │ 3.67435      1.0      0.0      0.0      0.0      1.0      1.0      0.0
    4 │ 2.66883      1.0      0.0      1.0      0.0      0.0      0.0      0.0
    5 │ 2.44569      0.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
    6 │ 2.80236      1.0      0.0      0.0      1.0      0.0      0.0      1.0
    7 │ 3.55435      0.0      0.0      1.0      0.0      0.0      0.0      0.0
    8 │ 2.62004      1.0      0.0      0.0      1.0      0.0      1.0      0.0
    9 │ 3.71028      0.0      0.0      0.0      1.0      0.0      0.0      1.0 ⋯
   10 │ 3.64966      1.0      0.0      0.0      0.0      1.0      1.0      0.0
   11 │ 3.07429      1.0      0.0      0.0      1.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 4111 │ 2.59984      1.0      0.0      1.0      0.0      0.0      0.0      1.0
 4112 │ 3.13883      1.0      0.0      0.0      0.0      1.0      0.0      0.0 ⋯
 4113 │ 2.78207      0.0      0.0      0.0      1.0      0.0      0.0      1.0
 4114 │ 3.36198      1.0      0.0      0.0      0.0      1.0      1.0      0.0
 4115 │ 3.29298      0.0      0.0      0.0      0.0      1.0      0.0      0.0
 4116 │ 3.21888      1.0      0.0      0.0      1.0      0.0      0.0      1.0 ⋯
 4117 │ 3.28193      1.0      0.0      1.0      0.0      0.0      0.0      0.0
 4118 │ 3.27497      0.0      0.0      0.0      0.0      1.0      1.0      0.0
 4119 │ 2.26835      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 4120 │ 3.64966      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
                                              1374 columns and 4099 rows omitted

Modelo Básico - Prueba:


1030×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.7052       1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │ 3.05182      0.0      0.0      0.0      1.0      0.0      1.0      0.0
    3 │ 3.14713      1.0      0.0      0.0      0.0      1.0      1.0      0.0
    4 │ 1.99766      1.0      0.0      1.0      0.0      0.0      0.0      0.0
    5 │ 2.7052       1.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
    6 │ 1.48944      0.0      0.0      0.0      1.0      0.0      1.0      0.0
    7 │ 2.78948      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    8 │ 3.59837      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    9 │ 3.21888      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
   10 │ 3.42652      1.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │ 2.48651      0.0      1.0      0.0      0.0      0.0      1.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 1021 │ 2.95651      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 1022 │ 4.20927      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
 1023 │ 4.43242      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 1024 │ 3.11778      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 1025 │ 4.11966      0.0      0.0      0.0      0.0      1.0      0.0      0.0
 1026 │ 2.62801      0.0      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
 1027 │ 3.36198      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 1028 │ 2.52573      1.0      1.0      0.0      0.0      0.0      1.0      0.0
 1029 │ 3.51613      1.0      1.0      0.0      0.0      0.0      0.0      0.0
 1030 │ 1.77994      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
                                              1374 columns and 1009 rows omitted

Modelo Flexible - Entrenamiento:


4120×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 3.51613      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
    2 │ 4.05512      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    3 │ 3.36198      1.0      0.0      0.0      1.0      0.0      0.0      1.0
    4 │ 2.66883      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    5 │ 1.82153      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
    6 │ 4.18029      0.0      0.0      0.0      0.0      1.0      1.0      0.0
    7 │ 2.59187      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │ 2.75806      1.0      0.0      0.0      1.0      0.0      0.0      0.0
    9 │ 2.62004      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
   10 │ 2.14558      1.0      0.0      1.0      0.0      0.0      1.0      0.0
   11 │ 3.58512      0.0      0.0      1.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 4111 │ 3.22618      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 4112 │ 2.94393      1.0      1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 4113 │ 2.56495      1.0      1.0      0.0      0.0      0.0      0.0      1.0
 4114 │ 2.91105      0.0      0.0      0.0      0.0      1.0      1.0      0.0
 4115 │ 2.91569      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 4116 │ 3.24419      0.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
 4117 │ 4.05512      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 4118 │ 2.158        1.0      0.0      1.0      0.0      0.0      1.0      0.0
 4119 │ 2.82298      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 4120 │ 2.9065       0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
                                              1374 columns and 4099 rows omitted

Modelo Flexible - Prueba:


1030×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.31466      1.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
    2 │ 3.5717       1.0      0.0      0.0      1.0      0.0      1.0      0.0
    3 │ 3.49651      1.0      0.0      0.0      1.0      0.0      1.0      0.0
    4 │ 2.82298      0.0      0.0      1.0      0.0      0.0      1.0      0.0
    5 │ 2.48651      1.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
    6 │ 3.46734      1.0      0.0      0.0      1.0      0.0      1.0      0.0
    7 │ 3.5717       1.0      0.0      0.0      1.0      0.0      1.0      0.0
    8 │ 3.5717       1.0      0.0      1.0      0.0      0.0      1.0      0.0
    9 │ 2.26336      0.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 2.82298      0.0      0.0      0.0      1.0      0.0      1.0      0.0
   11 │ 2.31215      0.0      0.0      0.0      0.0      0.0      1.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 1021 │ 2.41609      1.0      0.0      1.0      0.0      0.0      0.0      1.0
 1022 │ 2.28534      0.0      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
 1023 │ 1.83955      0.0      0.0      1.0      0.0      0.0      0.0      0.0
 1024 │ 2.66883      1.0      1.0      0.0      0.0      0.0      0.0      0.0
 1025 │ 1.46486      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 1026 │ 3.17966      1.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 1027 │ 4.50711      1.0      0.0      1.0      0.0      0.0      1.0      0.0
 1028 │ 3.98613      1.0      0.0      0.0      1.0      0.0      1.0      0.0
 1029 │ 4.34281      0.0      0.0      0.0      0.0      1.0      1.0      0.0
 1030 │ 2.36872      1.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
                                              1374 columns and 1009 rows omitted

Modelo Extra Flexible - Entrenamiento:


4120×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 3.11778      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
    2 │ 2.87855      1.0      0.0      0.0      0.0      1.0      0.0      0.0
    3 │ 1.96611      1.0      0.0      1.0      0.0      0.0      0.0      0.0
    4 │ 3.29298      0.0      0.0      1.0      0.0      0.0      1.0      0.0
    5 │ 3.36936      1.0      0.0      0.0      0.0      1.0      0.0      0.0 ⋯
    6 │ 3.6072       0.0      0.0      1.0      0.0      0.0      0.0      0.0
    7 │ 3.8526       1.0      0.0      0.0      0.0      1.0      0.0      1.0
    8 │ 2.66883      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │ 1.97568      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 4.04676      0.0      0.0      1.0      0.0      0.0      0.0      0.0
   11 │ 2.66883      1.0      0.0      1.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 4111 │ 3.18762      0.0      0.0      0.0      0.0      1.0      0.0      0.0
 4112 │ 3.68647      1.0      0.0      0.0      0.0      1.0      1.0      0.0 ⋯
 4113 │ 3.64966      1.0      0.0      0.0      1.0      0.0      1.0      0.0
 4114 │ 2.67988      1.0      0.0      0.0      1.0      0.0      0.0      0.0
 4115 │ 2.68307      1.0      0.0      0.0      1.0      0.0      0.0      0.0
 4116 │ 2.73059      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 4117 │ 2.19722      0.0      0.0      1.0      0.0      0.0      0.0      1.0
 4118 │ 2.93119      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 4119 │ 2.04022      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 4120 │ 3.06187      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
                                              1374 columns and 4099 rows omitted

Modelo Extra Flexible - Prueba:


1030×1382 DataFrame
  Row │ lwage    sex      hsg      scl      clg      ad       so       we      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ 2.59984      1.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
    2 │ 3.36198      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    3 │ 3.95981      0.0      0.0      0.0      1.0      0.0      1.0      0.0
    4 │ 4.05512      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    5 │ 2.82298      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │ 3.1797       0.0      1.0      0.0      0.0      0.0      1.0      0.0
    7 │ 2.26336      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │ 3.09627      1.0      0.0      0.0      1.0      0.0      0.0      0.0
    9 │ 3.07429      1.0      0.0      1.0      0.0      0.0      0.0      0.0 ⋯
   10 │ 2.31215      0.0      0.0      0.0      0.0      0.0      0.0      0.0
   11 │ 3.33495      0.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 1021 │ 3.13883      1.0      0.0      0.0      1.0      0.0      0.0      1.0
 1022 │ 3.36198      1.0      0.0      0.0      0.0      1.0      1.0      0.0 ⋯
 1023 │ 3.53188      1.0      0.0      0.0      1.0      0.0      0.0      0.0
 1024 │ 2.53666      1.0      0.0      1.0      0.0      0.0      0.0      0.0
 1025 │ 4.46059      1.0      0.0      0.0      0.0      1.0      0.0      1.0
 1026 │ 2.56347      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
 1027 │ 3.23095      1.0      1.0      0.0      0.0      0.0      0.0      1.0
 1028 │ 2.95651      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 1029 │ 2.82298      1.0      0.0      1.0      0.0      0.0      0.0      0.0
 1030 │ 3.32242      0.0      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
                                              1374 columns and 1009 rows omitted

10. Estimate all three models

In [59]:
# Se define una función para crear la fórmula dinámicamente usando todas las características excepto la target para crear un modelo que considere todas las otras variables
function create_formula(data::DataFrame, target::Symbol)
    # Ensure the target variable is not included in the features
    data_features = select(data, Not(target))  # Remove the target column from the DataFrame
    
    # Get all column names except the target
    features = names(data_features)  # Now, features will only contain independent variables
    
    # Convert the feature names to symbols
    features_symbols = Symbol.(features)
    
    # Create the formula using the terms as symbols
    return Term(target) ~ sum(Term.(features_symbols))
end

create_formula (generic function with 1 method)

In [60]:
# Mostrar un resumen de los resultados para cada modelo
function display_summary(model, model_name)
    println("Summary of OLS results for ", model_name)
    println(coeftable(model))  # Muestra la tabla de coeficientes
    println("R-squared: ", r2(model))  # Muestra el valor de R-cuadrado
    println("\n")
end

display_summary (generic function with 1 method)

basic

In [61]:
#fórmula del modelo 
formula_basico = create_formula(data_basico, :lwage)
ols_basico = lm(formula_basico, data_basico)
display_summary(ols_basico, "Basic Model")

Summary of OLS results for Basic Model
─────────────────────────────────────────────────────────────────────────────────────────────
                          Coef.    Std. Error       t  Pr(>|t|)      Lower 95%      Upper 95%
─────────────────────────────────────────────────────────────────────────────────────────────
(Intercept)         1.31601        1.19881       1.10    0.2724   -1.03427        3.66629
sex                -0.060786       0.01598      -3.80    0.0001   -0.0921149     -0.0294571
hsg                 1.83088        1.26706       1.44    0.1485   -0.653191       4.31496
scl                 1.95199        1.24582       1.57    0.1172   -0.490444       4.39442
clg                 2.22455        1.22216       1.82    0.0688   -0.171512       4.62061
ad                  2.23298        1.21239       1.84    0.0656   -0.143907       4.60987
so                  0.171589       0.209842      0.82    0.4136   -0.239807       0.582985
we                  0.131239       0.22322    

flexible

In [62]:
formula_flexible = create_formula(data_flexible, :lwage)
ols_flexible = lm(formula_flexible, data_flexible)
display_summary(ols_flexible, "Flexible Model")

Summary of OLS results for Flexible Model
─────────────────────────────────────────────────────────────────────────────────────────────
                          Coef.    Std. Error       t  Pr(>|t|)      Lower 95%      Upper 95%
─────────────────────────────────────────────────────────────────────────────────────────────
(Intercept)         1.31601        1.19881       1.10    0.2724   -1.03427        3.66629
sex                -0.060786       0.01598      -3.80    0.0001   -0.0921149     -0.0294571
hsg                 1.83088        1.26706       1.44    0.1485   -0.653191       4.31496
scl                 1.95199        1.24582       1.57    0.1172   -0.490444       4.39442
clg                 2.22455        1.22216       1.82    0.0688   -0.171512       4.62061
ad                  2.23298        1.21239       1.84    0.0656   -0.143907       4.60987
so                  0.171589       0.209842      0.82    0.4136   -0.239807       0.582985
we                  0.131239       0.22322 

extra-flexible model

In [63]:
formula_extra_flexible = create_formula(data_extra_flexible, :lwage)
ols_extra_flexible = lm(formula_extra_flexible, data_extra_flexible)
display_summary(ols_extra_flexible, "Extra Flexible Model")

Summary of OLS results for Extra Flexible Model
─────────────────────────────────────────────────────────────────────────────────────────────
                          Coef.    Std. Error       t  Pr(>|t|)      Lower 95%      Upper 95%
─────────────────────────────────────────────────────────────────────────────────────────────
(Intercept)         1.31601        1.19881       1.10    0.2724   -1.03427        3.66629
sex                -0.060786       0.01598      -3.80    0.0001   -0.0921149     -0.0294571
hsg                 1.83088        1.26706       1.44    0.1485   -0.653191       4.31496
scl                 1.95199        1.24582       1.57    0.1172   -0.490444       4.39442
clg                 2.22455        1.22216       1.82    0.0688   -0.171512       4.62061
ad                  2.23298        1.21239       1.84    0.0656   -0.143907       4.60987
so                  0.171589       0.209842      0.82    0.4136   -0.239807       0.582985
we                  0.131239       0.

11. Report the MSE and the R2 for both samples of each model, and the adjusted R2 for the training sample. Do we see overfitting? Which model has the best out-of-sample performance? For reference, you should report the following:

In [64]:
# Function to calculate Mean Squared Error (MSE)
function compute_mse(model, data)
    # Predict the values using the model
    predictions = predict(model, data)
    # Calculate the residuals
    residuals = data.lwage .- predictions
    # Compute MSE
    mse = mean(residuals .^ 2)
    return mse
end

# Function to compute Adjusted R-squared
function adjusted_r2(model, data)
    r2_val = r2(model)  # Compute R-squared
    n = nrow(data)  # Number of observations
    p = length(coef(model)) - 1  # Number of predictors
    adj_r2 = 1 - (1 - r2_val) * (n - 1) / (n - p - 1)  # Adjusted R-squared formula
    return adj_r2
end

# Compute metrics for each model
function report_metrics(formula, train_data, test_data, model_name)
    # Fit the model using the training data
    model_train = lm(formula, train_data)

    # Compute MSE for training and test samples
    mse_train = compute_mse(model_train, train_data)
    mse_test = compute_mse(model_train, test_data)

    # Compute R-squared for training and test samples
    r2_train = r2(model_train)
    r2_test = r2(lm(formula, test_data))

    # Compute Adjusted R-squared for training sample only
    adj_r2_train = adjusted_r2(model_train, train_data)

    # Print the results
    println("Metrics for ", model_name, ":")
    println("MSE (Training Sample): ", mse_train)
    println("MSE (Test Sample): ", mse_test)
    println("R-squared (Training Sample): ", r2_train)
    println("R-squared (Test Sample): ", r2_test)
    println("Adjusted R-squared (Training Sample): ", adj_r2_train)
    println("\n")
end

# Assuming the formulas have been created correctly
# Call the function to report metrics for each model
report_metrics(formula_basico, train_basico, test_basico, "Model 1")
report_metrics(formula_flexible, train_flexible, test_flexible, "Model 2")
report_metrics(formula_extra_flexible, train_extra_flexible, test_extra_flexible, "Model 3")

Metrics for Model 1:
MSE (Training Sample): 0.1680154846509501
MSE (Test Sample): 0.2952714675297894
R-squared (Training Sample): 0.4813767720110844
R-squared (Test Sample): 0.6489032760514655
Adjusted R-squared (Training Sample): 0.21979215628694537


Metrics for Model 2:
MSE (Training Sample): 0.1747749260893782
MSE (Test Sample): 0.2718895651329484
R-squared (Training Sample): 0.47040819497289543
R-squared (Test Sample): 0.7267879087341038
Adjusted R-squared (Training Sample): 0.20329121807646333


Metrics for Model 3:
MSE (Training Sample): 0.17227134748523415
MSE (Test Sample): 0.3988451540431668
R-squared (Training Sample): 0.46485252094508556
R-squared (Test Sample): 0.737924515722537
Adjusted R-squared (Training Sample): 0.19493335784251553




Interpretación: 
- En términos generales, los tres modelos presentan métricas de predicción más altas en el conjunto de entrenamiento, en comparación cuando se evalúan en el conjunto de prueba. 
- En el conjunto de entrenamiento, se puede observar que el modelo flexible tiene un mayor r cuadrado(0.327) en comparación con el modelo básico y el modelo más flexible. No obstante, cuando utilizamos el r cuadrado ajustado, el cual incluye una penalidad por la inclusión de variables irrelevantes, podemos observar que el modelo básico es el que posee la mayor capacidad predictiva (0.361). 
- Los resultados de evaluación en el conjunto de entrenamiento no se pueden replicar en el conjunto de prueba, pues allí el modelo que posee una mayor capacidad predictiva es el modelo más flexible (0.246)
- Se puede identificar un  ligero caso de overfitting con el segundo modelo, el que posee  el mayor número de variables (110), debido a que cuando se evalúa su capacidad predictiva en el conjunto de prueba, sus indicadores de evaluación decaen ligeramente (MSE_Train: 0.226, MSE_Test: 0.1881), (R2_Train: 0.326, R2_Test: 0.310), (Rajustado_Train: 0.245, Rajustado_Test: 0.227)